src/resolver.ts

```diff
@@ -3130,6 +3131,46 @@ export class Resolver extends DiagnosticEmitter {
               }
             }
             if (!fieldType) break; // did report above
+            if (existingField !== null) {
+              // visibility checks
+              let thisFieldIsPublic = fieldPrototype.isPublic;
+              let existingFieldIsPublic = existingField.isPublic;
+              let baseClass = <Class>base;
+
+              if (thisFieldIsPublic && !existingFieldIsPublic) {
+                this.errorRelated(
+                  DiagnosticCode.Property_0_is_private_in_type_1_but_not_in_type_2,
+                  fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                  fieldPrototype.name, baseClass.internalName, instance.internalName
+                );
+              } else if (!thisFieldIsPublic && existingFieldIsPublic) {
+                this.errorRelated(
+                  DiagnosticCode.Property_0_is_private_in_type_1_but_not_in_type_2,
+                  fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                  fieldPrototype.name, instance.internalName, baseClass.internalName
+                );
+              } else if (!thisFieldIsPublic && !existingFieldIsPublic) {
+                this.errorRelated(
+                  DiagnosticCode.Types_have_separate_declarations_of_a_private_property_0,
+                  fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                  fieldPrototype.name
+                );
+              } 
+              // assignability checks
+              else if (fieldType.equals(existingField.type) && fieldPrototype.initializerNode === null && !fieldPrototype.is(CommonFlags.DECLARE)) {
+                this.errorRelated(
+                  DiagnosticCode.Property_0_will_overwrite_the_base_property_in_1_If_this_is_intentional_add_an_initializer_Otherwise_add_a_declare_modifier_or_remove_the_redundant_declaration,
+                  fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                  fieldPrototype.name, baseClass.internalName
+                );
+              } else if (!fieldType.isStrictlyAssignableTo(existingField.type)) {
+                this.errorRelated(
+                  DiagnosticCode.Property_0_in_type_1_is_not_assignable_to_the_same_property_in_base_type_2,
+                  fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                  fieldPrototype.name, instance.internalName, baseClass.internalName
+                );
+              }
+            }
```

Comment:
Appears that below here, a new field with the same name and a different memory offset is created, but I think the memory offset should be reused if it is compatible?

Reply:
that's what I thought, but in the tests it appears as if the class has the proper size.. maybe I'm using the wrong functions in the test?

Reply:
The way to test this would be
```ts
class A {
  public foo: i32;
}
class B extends A {
  public foo: i32;
}
assert(offsetof<A>("foo") == 0);
assert(offsetof<B>("foo") == 0); // this one should fail if my assumption above is correct
```

Reply:
> if it is compatible

this just means if it's assignable right? or is there another check I should put in?

Reply:
Right, the sub type must be strictly assignable to the super type, that is something like `i32` that can technically assign to `i64` must be disallowed because it's not the right width in memory. Iirc you used `isStrictlyAssignable` there, which should be correct.